components/web_radio/web_radio.c

```diff
@@ -24,52 +24,122 @@
 #include "playlist.h"
 
 #define TAG "web_radio"
+#define HDR_KV_BUFF_LEN 128
 
 typedef enum
 {
-    HDR_CONTENT_TYPE = 1
+    HDR_CONTENT_TYPE = 1,
+    HDR_ICY_METAINT = 2
 } header_field_t;
 
-static header_field_t curr_header_field = 0;
-static content_type_t content_type = 0;
-static bool headers_complete = false;
+typedef enum {
+    HDR_UNDEF,
+    HDR_KEY,
+    HDR_VALUE,
+    HDR_COMPLETE
+} header_processing_state;
+
+header_field_t curr_header_field = 0;
+content_type_t content_type = 0;
+
+header_processing_state hdr_state= HDR_UNDEF;
+char hdr_key_buff[HDR_KV_BUFF_LEN] = {0};
+char hdr_value_buff[HDR_KV_BUFF_LEN] = {0};
+int hdr_key_buff_len = 0;
+int hdr_value_buff_len = 0;
+
+void process_header_key(const char *at, size_t length);
+void process_header_key_complete();
+void process_header_value(const char *at, size_t length);
+void process_header_value_complete();
+
+void process_header_key(const char *at, size_t length) {
+    if (hdr_state == HDR_VALUE) {
+        process_header_value_complete();
+    }
+    hdr_state = HDR_KEY;
 
-static int on_header_field_cb(http_parser *parser, const char *at, size_t length)
-{
-    // convert to lower case
-    unsigned char *c = (unsigned char *) at;
-    for (; *c; ++c)
-        *c = tolower(*c);
```

Comment:
is tolower() unnecessary?

Reply:
The `tolower` was intended to be used to compare header keys as some servers return them first character capitalized. Instead here it was causing the entire buffer to be converted to lower case. I did not want the metadata to be lowered, so I moved the comparison to `on_header_field_cb` which uses the `tolower` method.